src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -53,14 +94,32 @@ public ManagedChannel getChannel(String addressStr, HostMapping hostMapping) {
           }
           // Channel should be lazy without actual connection until first call
           // So a coarse grain lock is ok here
-          return ManagedChannelBuilder.forAddress(mappedAddr.getHost(), mappedAddr.getPort())
-              .maxInboundMessageSize(maxFrameSize)
-              .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
-              .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
-              .keepAliveWithoutCalls(true)
-              .usePlaintext(true)
-              .idleTimeout(60, TimeUnit.SECONDS)
-              .build();
+          if (sslContextBuilder == null) {
+            return ManagedChannelBuilder.forAddress(mappedAddr.getHost(), mappedAddr.getPort())
+                .maxInboundMessageSize(maxFrameSize)
+                .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
+                .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
+                .keepAliveWithoutCalls(true)
+                .usePlaintext(true)
+                .idleTimeout(60, TimeUnit.SECONDS)
+                .build();
+          } else {
+            SslContext sslContext = null;
+            try {
+              sslContext = sslContextBuilder.build();
+            } catch (SSLException e) {
+              logger.error("create ssl context failed!", e);
+              return null;
+            }
+            return NettyChannelBuilder.forAddress(mappedAddr.getHost(), mappedAddr.getPort())
+                .maxInboundMessageSize(maxFrameSize)
+                .keepAliveTime(keepaliveTime, TimeUnit.SECONDS)
+                .keepAliveTimeout(keepaliveTimeout, TimeUnit.SECONDS)
+                .keepAliveWithoutCalls(true)
+                .sslContext(sslContext)
+                .idleTimeout(60, TimeUnit.SECONDS)
+                .build();
```

Comment:
Is it possible to DRY?

Reply:
not support yet

Reply:
I think it is possible, see the example 11 in https://www.programcreek.com/java-api-examples/?class=io.grpc.netty.NettyChannelBuilder&method=sslContext

Reply:
fixed  @andylokandy PTAL